src/extrema_set.rs

```diff
@@ -0,0 +1,42 @@
+/// Implementation guts for `min_set`, `min_set_by`, and `min_set_by_key`.
+pub fn min_set_impl<I, K, F, L>(mut it: I,
+                                mut key_for: F,
+                                mut lt: L) -> Option<Vec<I::Item>>
+    where I: Iterator,
+          F: FnMut(&I::Item) -> K,
+          L: FnMut(&I::Item, &I::Item, &K, &K) -> bool,
+{
+    let (mut result, mut current_key) = match it.next() {
+        None => return None,
+        Some(element) => {
+            let key = key_for(&element);
+            (vec![element], key)
+        }
+    };
+
+    for element in it {
+        let key = key_for(&element);
+        if lt(&element, &result[0], &key, &current_key) {
```

Comment:
Would it make sense to replace `lt` (returning `bool`) by `cmp` (returning `Ordering`) so that we could avoid the second call to `lt` in line 23?

Reply:
That should work nicely.

As it is, I essentially copied the structure from the `minimax` set of functions, where I see now that multiple comparisons are also made.

I'm happy to make the change, but will probably not have the time to do so in the next couple of weeks (currently on vacation).

Reply:
Ah, I see. That didn't occur to me when I looked at `minmax` last time. I could, however, imagine that things are a bit more complicated there because it needs to look for minimum and maximum simultaneously.

Maybe I should go back and have a look at `min_max`.

Reply:
I had a look at `minmax` and think that it needs to do more stuff since it searches for both the minimum and the maximum. Here, I think we can use less calls to compare elements.